File: frontier-model-on-trivial-workload/node/openai-model-rightsizing-audit.mjs
/**
 * Report OpenAI models that are larger than the work they are doing.
 *
 * Read only. GET requests and nothing else: OPENAI_ADMIN_KEY must be an
 * organization admin key with read scopes, because every /v1/organization
 * endpoint rejects a project key outright. The repair is printed, never
 * performed.
 */
const API = 'https://api.openai.com/v1';

// Substrings that mean "this is already the small sibling".
const SMALL_MARKERS = ['mini', 'nano', 'small', 'lite', 'embedding', 'moderation'];

// The families worth right-sizing, each mapped to the cheaper sibling that
// answers the same shape of question. A table rather than a string rule,
// because a wrong suggestion here is worse than no suggestion.
const SIBLINGS = [
  ['gpt-5', 'gpt-5-mini'],
  ['gpt-4.1', 'gpt-4.1-mini'],
  ['gpt-4o', 'gpt-4o-mini'],
  ['o3', 'o4-mini'],
  ['o1', 'o4-mini'],
];

const FINDINGS = ['oversized'];

/**
 * Classify a model id. Pure, and deliberately conservative: "unknown" is not a
 * finding, because a model this table has never heard of is one this script has
 * no business advising on.
 */
export function tier(model) {
  const name = String(model ?? '').trim().toLowerCase();
  if (!name) return 'unknown';
  if (name.startsWith('ft:')) return 'custom';
  if (SMALL_MARKERS.some((m) => name.includes(m))) return 'small';
  for (const [family] of SIBLINGS) if (name.startsWith(family)) return 'premium';
  return 'unknown';
}

/** The cheaper model answering the same shape of question, or null. Pure. */
export function sibling(model) {
  const name = String(model ?? '').trim().toLowerCase();
  if (tier(name) !== 'premium') return null;
  for (const [family, cheaper] of SIBLINGS) {
    if (name.startsWith(family)) return cheaper;
  }
  return null;
}

/**
 * Sum the daily buckets into one row per model. Pure.
 *
 * Folding before dividing matters: a mean taken per bucket and then averaged
 * weights a quiet Sunday as heavily as a Tuesday.
 */
export function fold(pages) {
  const out = new Map();
  for (const page of pages) {
    for (const bucket of page.data ?? []) {
      for (const result of bucket.results ?? []) {
        const model = String(result.model ?? '').trim();
        if (!model) continue;
        if (!out.has(model)) {
          out.set(model, { requests: 0, input: 0, output: 0, projects: new Set() });
        }
        const row = out.get(model);
        for (const [field, key] of [['num_model_requests', 'requests'],
                                    ['input_tokens', 'input'],
                                    ['output_tokens', 'output']]) {
          const n = Number(result[field] ?? 0);
          if (Number.isFinite(n)) row[key] += Math.trunc(n);
        }
        if (result.project_id) row.projects.add(String(result.project_id));
      }
    }
  }
  const folded = {};
  for (const [model, row] of out) {
    folded[model] = { ...row, projects: [...row.projects].sort() };
  }
  return folded;
}

/**
 * Classify one folded model row. Pure. Returns [state, detail].
 * Short answers over enormous prompts are separated out, because the money
 * there is on the input side and swapping the model saves almost none of it.
 */
export function verdict(model, row, minRequests = 500, trivialOutput = 50,
                        longInput = 20000) {
  const requestsMade = Number(row.requests ?? 0);
  if (!Number.isFinite(requestsMade)) {
    return ['unreadable',
      'num_model_requests did not sum to a number, so there is no denominator ' +
      'and no ratio to read'];
  }
  if (requestsMade <= 0) {
    return ['unreadable', '0 request(s) in the window, so there is nothing to divide by'];
  }
  if (requestsMade < minRequests) {
    return ['low-volume',
      `${requestsMade} request(s) in the window, under the floor of ` +
      `${minRequests}. A mean taken over this few calls is noise, not a shape.`];
  }

  const outPer = Number(row.output ?? 0) / requestsMade;
  const inPer = Number(row.input ?? 0) / requestsMade;
  const shape = `${requestsMade} request(s), mean output ${outPer.toFixed(0)} ` +
                `token(s), mean input ${inPer.toFixed(0)} token(s)`;

  const kind = tier(model);
  if (kind === 'custom') {
    return ['custom-model',
      `${shape}. This is a fine-tune, and its size is inherited from the base ` +
      'model rather than chosen here.'];
  }
  if (kind === 'small') {
    return ['right-sized', `${shape}. Already the cheap sibling for its family.`];
  }
  if (kind !== 'premium') {
    return ['unknown-model',
      `${shape}. No cheaper sibling is known for this model id, so this script ` +
      'has no recommendation to make about it.'];
  }

  if (outPer >= trivialOutput) {
    return ['deliberative',
      `${shape}. The answers are long enough that the model is plausibly doing ` +
      'the work it was chosen for.'];
  }
  if (inPer >= longInput) {
    return ['input-bound',
      `${shape}. Short answers over very large prompts. The bill here is input, ` +
      'not model tier, so caching the prefix will save more than downgrading ' +
      'the model.'];
  }
  return ['oversized',
    `${shape}. A premium model returning answers this short is answering ` +
    'questions a cheaper sibling would answer identically.'];
}

/**
 * Can this project still reach this model? Pure. An unconstrained project is
 * the durable half of the finding: without a restriction the expensive model
 * comes back the next time somebody copies a snippet from the quickstart.
 */
export function permissionsState(perms, model) {
  if (perms === null || typeof perms !== 'object' || Array.isArray(perms)) {
    return 'unreadable';
  }
  const mode = String(perms.mode ?? '').trim().toLowerCase();
  const ids = (Array.isArray(perms.model_ids) ? perms.model_ids : [])
    .map((i) => String(i).trim().toLowerCase());
  const name = String(model ?? '').trim().toLowerCase();

  if (mode === 'allow_list') {
    if (ids.length === 0) return 'blocked';
    return ids.includes(name) ? 'allowed' : 'blocked';
  }
  if (mode === 'deny_list') {
    if (ids.length === 0) return 'unconstrained';
    return ids.includes(name) ? 'blocked' : 'allowed';
  }
  return 'unreadable';
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) {
    if (Array.isArray(v)) for (const item of v) url.searchParams.append(k, String(item));
    else if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from OpenAI: OPENAI_ADMIN_KEY must be an organization ' +
                    'admin key, not a project key');
  }
  if (res.status === 403) {
    throw new Error('403 from OpenAI: the key is not authorised for ' +
                    '/v1/organization. A project key cannot read usage.');
  }
  if (!res.ok) throw new Error(`${res.status} from ${path}`);
  return res.json();
}

async function usagePages(key, startTime, days, maxPages = 20) {
  const pages = [];
  let params = {
    start_time: startTime, bucket_width: '1d', limit: days,
    group_by: ['model', 'project_id'],
  };
  for (let i = 0; i < maxPages; i += 1) {
    const page = await get(key, '/organization/usage/completions', params);
    pages.push(page);
    if (!page.next_page) break;
    params = { ...params, page: page.next_page };
  }
  return pages;
}

async function spendByLineItem(key, startTime) {
  const out = {};
  const page = await get(key, '/organization/costs',
    { start_time: startTime, limit: 31, group_by: 'line_item' });
  for (const bucket of page.data ?? []) {
    for (const result of bucket.results ?? []) {
      const item = String(result.line_item ?? '');
      const amount = Number(result.amount?.value ?? 0);
      if (Number.isFinite(amount)) out[item] = (out[item] ?? 0) + amount;
    }
  }
  return out;
}

/**
 * Spend on exactly this model, from the cost report's line items. Pure.
 * Substring matching is not good enough: "gpt-5" occurs inside "gpt-5-mini,
 * input tokens" and inside a fine-tune id built on it, and quoting either as
 * the premium model's spend overstates the saving in the one line a reader is
 * going to act on. Model ids only contain letters, digits, dots and dashes, so
 * the escape is a pair of character classes rather than a backslash dance.
 */
export function spendFor(model, spend) {
  const name = String(model ?? '').trim().toLowerCase();
  if (!name) return 0;
  const escaped = name.replace(/[.-]/g, (c) => `[${c}]`);
  const pattern = new RegExp(`(?<![-a-z0-9.:])${escaped}(?![-a-z0-9.])`);
  let total = 0;
  for (const [item, amount] of Object.entries(spend ?? {})) {
    if (pattern.test(String(item).toLowerCase())) {
      const value = Number(amount);
      if (Number.isFinite(value)) total += value;
    }
  }
  return total;
}

async function main() {
  const key = (process.env.OPENAI_ADMIN_KEY || "dummy-openai-admin-key");
  if (!key) {
    console.error('set OPENAI_ADMIN_KEY (an organization admin key with read scopes)');
    process.exitCode = 2;
    return;
  }

  const days = Number((process.env.DAYS || "dummy-days") ?? 14);
  const minRequests = Number((process.env.MIN_REQUESTS || "dummy-min-requests") ?? 500);
  const trivialOutput = Number((process.env.TRIVIAL_OUTPUT || "dummy-trivial-output") ?? 50);
  const showAll = process.argv.includes('--show-all');

  const now = Math.floor(Date.now() / 1000);
  const rows = fold(await usagePages(key, now - days * 86400, days));
  const spend = await spendByLineItem(key, now - 30 * 86400);

  let checked = 0;
  let bad = 0;
  for (const model of Object.keys(rows).sort()) {
    const row = rows[model];
    const [state, detail] = verdict(model, row, minRequests, trivialOutput);
    checked += 1;
    const line = `${state.padEnd(14)} ${model.padEnd(16)} ${detail}`;

    if (FINDINGS.includes(state)) {
      bad += 1;
      console.warn(line);
      const cheaper = sibling(model);
      console.warn(`  repair: ${cheaper} answers this shape of question; 30d ` +
                   `spend on ${model} was $${spendFor(model, spend).toFixed(2)}`);
      for (const project of row.projects) {
        const perms = await get(key,
          `/organization/projects/${project}/model_permissions`);
        const where = permissionsState(perms, model);
        if (where === 'unconstrained') {
          console.warn(`  repair: project ${project} is unconstrained. To make ` +
            `the change durable, set model_permissions to mode allow_list with ` +
            `model_ids ['${cheaper}'] so the expensive model cannot come back.`);
        } else {
          console.warn(`  note: project ${project} model_permissions say ${where}`);
        }
      }
    } else if (state === 'input-bound') {
      console.warn(line);
      console.warn('  repair: read the prompt-caching note before changing the ' +
                   'model. A stable prefix at this size is the bill.');
    } else if (state === 'unreadable') {
      console.warn(line);
    } else if (showAll) {
      console.log(line);
    }
  }

  console.log(`${checked} model(s) checked, ${bad} finding(s)`);
  process.exitCode = bad ? 1 : 0;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: frontier-model-on-trivial-workload/python/openai_model_rightsizing_audit.py
"""Report OpenAI models that are larger than the work they are doing.

Read only. GET requests and nothing else: OPENAI_ADMIN_KEY must be an
organization admin key (sk-admin-...) with read scopes, because every
/v1/organization endpoint rejects a project key outright.

The repair is printed, never performed. Which model serves production traffic
is a deploy, and restricting a project's model permissions changes what your
colleagues are allowed to call. Neither belongs to an audit script.
"""
import argparse
import datetime as dt
import logging
import os
import re
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("openai_model_rightsizing_audit")

API = "https://api.openai.com/v1"

# Substrings that mean "this is already the small sibling". Matched on the model
# id because there is no field on the usage result that says how big a model is.
SMALL_MARKERS = ("mini", "nano", "small", "lite", "embedding", "moderation")

# The families worth right-sizing, in the order they are tested. Each maps to the
# cheaper sibling that answers the same shape of question. Kept as a table rather
# than a string rule because "gpt-5" -> "gpt-5-mini" is a naming convention, not
# a guarantee, and a wrong suggestion here is worse than none.
SIBLINGS = (
    ("gpt-5", "gpt-5-mini"),
    ("gpt-4.1", "gpt-4.1-mini"),
    ("gpt-4o", "gpt-4o-mini"),
    ("o3", "o4-mini"),
    ("o1", "o4-mini"),
)

FINDINGS = ("oversized",)


def tier(model):
    """Classify a model id. Pure, and deliberately conservative.

    Returns "custom" for a fine-tune, "small" for a model that is already the
    cheap sibling, "premium" for a family with a cheaper sibling to move to, and
    "unknown" for everything else. Unknown is not a finding: a model this table
    has never heard of is a model this script has no business advising on.
    """
    name = str(model or "").strip().lower()
    if not name:
        return "unknown"
    if name.startswith("ft:"):
        return "custom"
    if any(marker in name for marker in SMALL_MARKERS):
        return "small"
    for family, _cheaper in SIBLINGS:
        if name.startswith(family):
            return "premium"
    return "unknown"


def sibling(model):
    """The cheaper model that answers the same shape of question, or None. Pure."""
    name = str(model or "").strip().lower()
    if tier(name) != "premium":
        return None
    for family, cheaper in SIBLINGS:
        if name.startswith(family):
            return cheaper
    return None


def fold(pages):
    """Sum the daily buckets into one row per model. Pure.

    Folding before dividing matters: a mean taken per bucket and then averaged
    weights a quiet Sunday exactly as heavily as a Tuesday, which is how a model
    that is busy on weekdays acquires a flattering output-per-request number.

    project_ids are collected as a sorted list so the caller knows which projects
    to ask about model permissions, and so two runs print the same order.
    """
    out = {}
    for page in pages:
        for bucket in page.get("data") or []:
            for result in bucket.get("results") or []:
                model = str(result.get("model") or "").strip()
                if not model:
                    continue
                row = out.setdefault(model, {"requests": 0, "input": 0,
                                             "output": 0, "projects": set()})
                for field, key in (("num_model_requests", "requests"),
                                   ("input_tokens", "input"),
                                   ("output_tokens", "output")):
                    try:
                        row[key] += int(result.get(field) or 0)
                    except (TypeError, ValueError):
                        pass
                project = result.get("project_id")
                if project:
                    row["projects"].add(str(project))
    return {m: {**row, "projects": sorted(row["projects"])} for m, row in out.items()}


def verdict(model, row, min_requests=500, trivial_output=50, long_input=20000):
    """Classify one folded model row. Pure. Returns (state, detail).

    The order is the argument. A model with too few calls has no shape to read.
    A model that is already small, or that this script does not recognise, is
    not advised on at all. Only then does the ratio decide, and short answers
    over enormous prompts are separated out because the money there is on the
    input side and swapping the model saves almost none of it.
    """
    try:
        requests_made = int(row.get("requests") or 0)
    except (TypeError, ValueError):
        return ("unreadable",
                "num_model_requests did not sum to an integer, so there is no "
                "denominator and no ratio to read")
    if requests_made <= 0:
        return ("unreadable",
                "0 request(s) in the window, so there is nothing to divide by")
    if requests_made < min_requests:
        return ("low-volume",
                "%d request(s) in the window, under the floor of %d. A mean "
                "taken over this few calls is noise, not a shape."
                % (requests_made, min_requests))

    out_per = (row.get("output") or 0) / float(requests_made)
    in_per = (row.get("input") or 0) / float(requests_made)
    shape = ("%d request(s), mean output %.0f token(s), mean input %.0f token(s)"
             % (requests_made, out_per, in_per))

    kind = tier(model)
    if kind == "custom":
        return ("custom-model",
                "%s. This is a fine-tune, and its size is inherited from the "
                "base model rather than chosen here." % shape)
    if kind == "small":
        return ("right-sized",
                "%s. Already the cheap sibling for its family." % shape)
    if kind != "premium":
        return ("unknown-model",
                "%s. No cheaper sibling is known for this model id, so this "
                "script has no recommendation to make about it." % shape)

    if out_per >= trivial_output:
        return ("deliberative",
                "%s. The answers are long enough that the model is plausibly "
                "doing the work it was chosen for." % shape)
    if in_per >= long_input:
        return ("input-bound",
                "%s. Short answers over very large prompts. The bill here is "
                "input, not model tier, so caching the prefix will save more "
                "than downgrading the model." % shape)
    return ("oversized",
            "%s. A premium model returning answers this short is answering "
            "questions a cheaper sibling would answer identically." % shape)


def permissions_state(perms, model):
    """Can this project still reach this model? Pure. Returns a state string.

    GET /v1/organization/projects/{id}/model_permissions returns a mode of
    allow_list or deny_list with a model_ids array. An unconstrained project is
    the durable half of the finding: without a restriction the expensive model
    comes back the next time somebody copies a snippet from the quickstart.
    """
    if not isinstance(perms, dict):
        return "unreadable"
    mode = str(perms.get("mode") or "").strip().lower()
    ids = perms.get("model_ids")
    if not isinstance(ids, list):
        ids = []
    ids = [str(i).strip().lower() for i in ids]
    name = str(model or "").strip().lower()

    if mode == "allow_list":
        if not ids:
            return "blocked"
        return "allowed" if name in ids else "blocked"
    if mode == "deny_list":
        if not ids:
            return "unconstrained"
        return "blocked" if name in ids else "allowed"
    return "unreadable"


def get(session, path, params=None):
    r = session.get(API + path, params=params or {}, timeout=60)
    if r.status_code == 401:
        raise SystemExit("401 from OpenAI: OPENAI_ADMIN_KEY must be an "
                         "organization admin key, not a project key")
    if r.status_code == 403:
        raise SystemExit("403 from OpenAI: the key is not authorised for "
                         "/v1/organization. A project key cannot read usage.")
    r.raise_for_status()
    return r.json()


def usage_pages(session, start_time, days, max_pages=20):
    """Walk the usage endpoint, which paginates on next_page."""
    params = {"start_time": start_time, "bucket_width": "1d", "limit": days,
              "group_by": ["model", "project_id"]}
    for _ in range(max_pages):
        page = get(session, "/organization/usage/completions", params)
        yield page
        cursor = page.get("next_page")
        if not cursor:
            return
        params = dict(params, page=cursor)


def spend_by_line_item(session, start_time):
    """Thirty days of spend, keyed by the cost report's line_item string."""
    out = {}
    page = get(session, "/organization/costs",
               {"start_time": start_time, "limit": 31, "group_by": "line_item"})
    for bucket in page.get("data") or []:
        for result in bucket.get("results") or []:
            item = str(result.get("line_item") or "")
            amount = (result.get("amount") or {}).get("value") or 0
            try:
                out[item] = out.get(item, 0.0) + float(amount)
            except (TypeError, ValueError):
                pass
    return out


def spend_for(model, spend):
    """Spend on exactly this model, from the cost report's line items. Pure.

    Substring matching is not good enough here. "gpt-5" occurs inside
    "gpt-5-mini, input tokens" and inside a fine-tune id built on it, and
    quoting either as the premium model's spend overstates the saving in the
    one line a reader is actually going to act on. So the match has to sit
    between boundaries: no letter, digit, dot, dash or colon on either side.
    """
    name = str(model or "").strip().lower()
    if not name:
        return 0.0
    pattern = re.compile(r"(?<![-a-z0-9.:])" + re.escape(name) + r"(?![-a-z0-9.])")
    total = 0.0
    for item, amount in (spend or {}).items():
        if pattern.search(str(item).lower()):
            try:
                total += float(amount)
            except (TypeError, ValueError):
                pass
    return total


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--days", type=int, default=14,
                    help="days of usage to fold (default 14)")
    ap.add_argument("--min-requests", type=int, default=500,
                    help="ignore models with fewer calls than this (default 500)")
    ap.add_argument("--trivial-output", type=int, default=50,
                    help="mean output tokens under which work is trivial (default 50)")
    ap.add_argument("--show-all", action="store_true",
                    help="also print models that are the right size")
    args = ap.parse_args()

    key = os.environ.get("OPENAI_ADMIN_KEY")
    if not key:
        log.error("set OPENAI_ADMIN_KEY (an organization admin key with read scopes)")
        return 2

    session = requests.Session()
    session.headers.update({"Authorization": "Bearer " + key})

    now = dt.datetime.now(dt.timezone.utc)
    usage_start = int((now - dt.timedelta(days=args.days)).timestamp())
    cost_start = int((now - dt.timedelta(days=30)).timestamp())

    rows = fold(usage_pages(session, usage_start, args.days))
    spend = spend_by_line_item(session, cost_start)

    checked = 0
    bad = 0
    for model in sorted(rows):
        row = rows[model]
        state, detail = verdict(model, row, args.min_requests, args.trivial_output)
        checked += 1
        line = "%-14s %-16s %s" % (state, model, detail)

        if state in FINDINGS:
            bad += 1
            log.warning(line)
            cheaper = sibling(model)
            money = spend_for(model, spend)
            log.warning("  repair: %s answers this shape of question; 30d spend "
                        "on %s was $%.2f", cheaper, model, money)
            for project in row["projects"]:
                perms = get(session,
                            "/organization/projects/%s/model_permissions" % project)
                where = permissions_state(perms, model)
                if where == "unconstrained":
                    log.warning("  repair: project %s is unconstrained. To make "
                                "the change durable, set model_permissions to "
                                "mode allow_list with model_ids [%r] so the "
                                "expensive model cannot come back.",
                                project, cheaper)
                else:
                    log.warning("  note: project %s model_permissions say %s",
                                project, where)
        elif state == "input-bound":
            log.warning(line)
            log.warning("  repair: read the prompt-caching note before changing "
                        "the model. A stable prefix at this size is the bill.")
        elif state in ("unreadable",):
            log.warning(line)
        elif args.show_all:
            log.info(line)

    log.info("%d model(s) checked, %d finding(s)", checked, bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())
